### PR TITLE
Support for recursive refs

### PR DESCRIPTION
**Pull request by analogue**
See #35 for context.

This PR adds support for recursive $refs. This is not something that can be broken up and added incrementally, hence the large diff. Basically, any read/write operations to ``swagger_spec.spec_dict`` will attempt to dereference a ``$ref`` if present. The ``RefResolver`` from the jsonschema library is used extensively to perform the dereferencing.

Highlights:
- jsonref no longer a dependency
- petstore spec in test-data updated to latest and greatest
- ~~swagger spec validation (not request/response validation) is temporarily disabled until swagger-spec-validator is also updated to support recursive refs~~
- ``bravado_core.spec.Spec.deref(..)`` and ``bravado_core.spec.post_process_spec(...)`` are temporary. They will use similiar functions that will reside in swagger-spec-validator.
- Minor doc updates along the way
- ``bravado_core.operation.build_params(..)`` fixed to support overriding same parameter at the path level
- Un-qualified refs such as ``{"$ref": "Category"}`` no longer supported  (follwing the lead of the petstore example)
- Automatic model tagging is only supported in the top-level spec file (determining fully qualified json pointer path is a little complicated with traversing non-root level refs)
- unit test suite passes

### CODE REVIEW

File: bravado_core/marshal.py [hunk 1]
```diff
@@ -22,11 +19,14 @@ def marshal_schema_object(swagger_spec, schema_object_spec, value):
...
     :type value: int, long, string, unicode, boolean, list, dict, Model type
+
     :return: marshaled value
     :rtype: int, long, string, unicode, boolean, list, dict
     :raises: SwaggerMappingError
     """
-    obj_type = schema_object_spec['type']
+    deref = swagger_spec.deref
+    schema_object_spec = deref(schema_object_spec)
+    obj_type = schema_object_spec.get('type')
```
Comment (prat0318) [inline bravado_core/marshal.py]: isn't 'type' a necessary key?

File: bravado_core/marshal.py [hunk 2]
```diff
@@ -3,17 +3,14 @@
...
 from bravado_core.model import is_model, MODEL_MARKER
-from bravado_core.schema import (
-    is_dict_like,
-    is_list_like,
-    SWAGGER_PRIMITIVES,
-    get_spec_for_prop
-)
+from bravado_core.schema import is_dict_like
+from bravado_core.schema import is_list_like
+from bravado_core.schema import SWAGGER_PRIMITIVES
```
Comment (prat0318) [inline bravado_core/marshal.py]: nit: ordering

File: bravado_core/model.py [hunk 3]
```diff
@@ -1,105 +1,73 @@
...
+    if len(path) < 2 or path[-2] != 'definitions':
         return
+    deref = swagger_spec.deref
+    model_name = key
+    model_spec = deref(container.get(key))
 
-    ref_target = jsonref_proxy.__reference__['$ref']
-    match = RE_MODEL_NAME.match(ref_target)
-    if match is None:
+    if deref(model_spec.get('type')) != 'object':
```
Comment (prat0318) [inline bravado_core/model.py]: Isn't mode_spec already all `deref`ed? do we need another `deref()`?

File: bravado_core/resource.py [hunk 4]
```diff
@@ -45,46 +45,48 @@ def build_resources(swagger_spec):
...
+        resources[tag] = Resource(tag, ops)
     return resources
 
 
 class Resource(object):
+    """A Swagger resource is associated with multiple operations.
+
+    :param name: resource name
+    :type name: str
+    :param operations: operations associated with this resource (by tag)
```
Comment (prat0318) [inline bravado_core/resource.py]: s/operations/ops

**Comment by prat0318** [comment]
there has to be all print cleanups. looks good otherwise. i haven't looked much in the tests though.

**Comment by analogue** [comment]
Blocked on https://github.com/Yelp/swagger_spec_validator/pull/41 getting merged and released

File: bravado_core/model.py (hunk 3 above)
Comment (analogue) [inline bravado_core/model.py]: ``deref`` only applies to the value passed in.

**Pull request merged**